File: golang/hackerrank/ModifiedKaprekarNumbers.c
// https://www.hackerrank.com/challenges/kaprekar-numbers/problem

#include <stdio.h>
#include <math.h>

#define SIZE 100000

int K[SIZE];

long
isKaprekar(int x)
{
    long int sum;
    long int a = 0;
    long int k = i;
    while(k>0) {
        k/=10;
        ++a;
    }
    long int p = x * x;
    long int g = pow(10,a);
    sum = p%g + p/g;
    return sum == x;
}

void
kaprekar() {
    long i, j, n, sq, dNo, mid, left, right, d[12];
   
    for(n = 1; n < SIZE; n++) {
        sq = n * n;
        
        dNo = 0;
        while(sq) {
            d[++dNo] = sq % 10;
            sq /= 10;
        }
        
        mid = dNo / 2;
        for(j = dNo, i = 1; i <= dNo/2; i++, j--) d[i] ^= d[j] ^= d[i] ^= d[j];
        
        for(left = j = 0, i = mid; i >= 1; i--, j++) {
            left += d[i] * pow(10, j);
        }
        for(right = j = 0, i = dNo; i >= mid + 1; i--, j++) {
            right += d[i] * pow(10, j);
        }

        if(left + right == n) K[n] = 1;
        else                  K[n] = 0;
    }
}

int main() {
    long int p,q; // r,f;
    int count = 0;
    scanf("%ld\n%ld",&p,&q);
    for( long int x=p; x<=q; x++ ) {
        if (isKaprekar(x)) {
            printf("%ld ",x);
            count++;
        }
    }
    if(count == 0)
        printf("INVALID RANGE");       
    return 0;
}


int main() {
   
    kaprekar();
    
    int i, p, q, flag;
    scanf(" %d %d", &p, &q);
    
    for(flag = 1, i = p; i <= q; i++) {
        if(K[i]) {
            printf("%d ", i);
            flag = 0;
        }
    }
    if(flag) printf("INVALID RANGE");
    
    return 0;
}

File: golang/hackerrank/ModifiedKaprekarNumbers.go
// https://www.hackerrank.com/challenges/kaprekar-numbers/problem

/*
A modified Kaprekar number is a positive whole number with a special property.
If you square it, then split the number into two integers and sum those integers,
you have the same value you started with.

Consider a positive whole number n, with d, digits.
We square n to arrive at a number that is either (2 x d) digits long or (2 x d) digits long.
Split the string representation of the square into two parts, l and r.
The right hand part r must be d digits long.
The left is the remaining substring.
Convert those two substrings back to integers, add them and see if you get.
*/

package main

import (
    "bufio"
    "fmt"
    "io"
    "os"
    "strconv"
    "strings"
)

func isKaprekar(x int32) (int32,bool) {
    number := strconv.Itoa(int(x)*int(x))
    n := len(number)
    digits := n/2
    s := number[0:digits]
    var left int64 = 0
    var right int64 = 0
    if len(s) > 0 {
        left,_ = strconv.ParseInt(s, 10, 64)
        // if err != nil {
        //     left = 0
        // }
    }
    s = number[digits:]
    if len(s) > 0 {
        right,_ = strconv.ParseInt(s, 10, 64)
        // if err != nil {
        //     right = 0
        // }
    }
    if int(left+right) == int(x) {
        return x, true
    }
    return 0, false
}
/*
 * Complete the 'kaprekarNumbers' function below.
 *
 * The function accepts following parameters:
 *  1. INTEGER p: the lower limit
 *  2. INTEGER q: the upper limit
 */

func kaprekarNumbers(p int32, q int32) {
    // Write your code here
    var kaprekar []string = make([]string,0)
    for x:=p; x<=q; x++ {
        if _,ok := isKaprekar(x); ok {
            kaprekar = append(kaprekar,strconv.Itoa(int(x)))
        }
    }
    if len(kaprekar) == 0 {
        fmt.Println("INVALID RANGE")
    } else {
        fmt.Println(strings.Join(kaprekar," "))
    }
    return
}

func main() {
    reader := bufio.NewReaderSize(os.Stdin, 16 * 1024 * 1024)

    pTemp, err := strconv.ParseInt(strings.TrimSpace(readLine(reader)), 10, 64)
    checkError(err)
    p := int32(pTemp)

    qTemp, err := strconv.ParseInt(strings.TrimSpace(readLine(reader)), 10, 64)
    checkError(err)
    q := int32(qTemp)

    kaprekarNumbers(p, q)
}

func readLine(reader *bufio.Reader) string {
    str, _, err := reader.ReadLine()
    if err == io.EOF {
        return ""
    }

    return strings.TrimRight(string(str), "\r\n")
}

func checkError(err error) {
    if err != nil {
        panic(err)
    }
}

